src/ImpactEvaluator.sol

```diff
@@ -88,45 +87,50 @@ contract ImpactEvaluator is AccessControl {
         Round storage round = openRounds[roundIndex];
         require(round.exists, "Open round does not exist");
 
-        for (uint i = 0; i < addresses.length; i++) {
-            round.addresses.push(addresses[i]);
-            round.scores.push(scores[i]);
-        }
+        uint sumOfScores = validateScores(scores, round.totalScores);
+        reward(addresses, scores);
+        round.totalScores += sumOfScores;
 
-        if (allScoresSubmitted(round.scores)) {
-            reward(round.addresses, round.scores);
-            cleanUpRound(round, roundIndex);
+        if (round.totalScores == MAX_SCORE) {
```

Comment:
Why can't we just delete the round after rewards have been processed? Not entirely getting the need for this score check? 

Also if we do need this score check? How likely is it that the scores will exactly line up with the `MAX_SCORE`?

Reply:
This is because of a vulnerability discovered in the audit: If a buggy or malfunctioning evaluation service submits too high scores, it can drain the whole contract balance in one go. By tracking the `.totalScores`, we ensure we never pay out more for one round.

I think eventually this could be made even safer by deploying one reward contract _per round_, which has a fixed balance and can't spend more even without manual checking.

Reply:
> Also if we do need this score check? How likely is it that the scores will exactly line up with the MAX_SCORE?

It's the job of the evaluation service to do this. Iiuc as long as we get the rounding right, we can make it always equal `MAX_SCORE`.

I was wondering though, if we mess this up then we'll have rounds that will persist forever, so I think we might need to add an admin method for ending rounds without `MAX_SCORE` scores.

Reply:
Thanks for explaining Julian, agree with you that the IE needs this logic. 

> I was wondering though, if we mess this up then we'll have rounds that will persist forever, so I think we might need to add an admin method for ending rounds without MAX_SCORE scores.

I do agree we definitely need a failsafe against rounding errors from the evaluation service. We already ran into conversion / rounding issues in our contracts ([relevant issue](https://github.com/filecoin-saturn/contracts/issues/67).

Reply:
That's useful to know, thanks! Added 🙏